File: src/Software Engineer Prepare Kit/34/solution.go
package main

type SinglyLinkedListNode struct {
	data int32
	next *SinglyLinkedListNode
}

type SinglyLinkedList struct {
	head *SinglyLinkedListNode
	tail *SinglyLinkedListNode
}

func (singlyLinkedList *SinglyLinkedList) InsertNodeIntoSinglyLinkedList(nodeData int32) {
	node := &SinglyLinkedListNode{
		next: nil,
		data: nodeData,
	}

	if singlyLinkedList.head == nil {
		singlyLinkedList.head = node
	} else {
		singlyLinkedList.tail.next = node
	}

	singlyLinkedList.tail = node
}

func ExtractAndAppendSponsoredNodes(head *SinglyLinkedListNode) *SinglyLinkedListNode {
	if head == nil || head.next == nil {
		return head
	}

	evens := head
	odds := head.next
	evens.next = nil

	dummyOdds := &SinglyLinkedListNode{
		next: odds,
	}

	// Link evens in reverse order and odds
	for dummyOdds.next != nil && dummyOdds.next.next != nil {
		dummyOddsNext := dummyOdds.next.next
		dummyOdds.next.next = dummyOdds.next.next.next

		dummyOddsNext.next = evens

		evens = dummyOddsNext
		dummyOdds = dummyOdds.next
	}

	// Attach odds tail with evens head
	if dummyOdds.next == nil {
		dummyOdds.next = evens
	} else {
		dummyOdds.next.next = evens
	}

	head = odds

	return head
}

File: src/Software Engineer Prepare Kit/34/solution.ts
class SinglyLinkedListNode {
	data: number
	next: SinglyLinkedListNode | null

	constructor(nodeData: number) {
		this.data = nodeData
		this.next = null
	}
}

class SinglyLinkedList {
	head: SinglyLinkedListNode | null
	tail: SinglyLinkedListNode | null

	constructor() {
		this.head = null
		this.tail = null
	}

	insertNode(nodeData: number): void {
		const node = new SinglyLinkedListNode(nodeData)

		if (this.head == null) {
			this.head = node
		} else {
			this.tail!.next = node
		}

		this.tail = node
	}
}

export default function extractAndAppendSponsoredNodes(head: SinglyLinkedListNode): SinglyLinkedListNode {
	if (!head || !head.next) return head

	let evens = head
	const odds = head.next
	head.next = null

	let dummyOdds = new SinglyLinkedListNode(0)
	dummyOdds.next = odds

	// Links evens in reverse order and odds
	while (dummyOdds.next !== null && dummyOdds.next.next !== null) {
		const dummyOddsNext = dummyOdds.next.next
		dummyOdds.next.next = dummyOdds.next.next.next

		dummyOddsNext.next = evens

		evens = dummyOddsNext
		dummyOdds = dummyOdds.next
	}

	// Attach odds tail with evens head
	if (!dummyOdds.next) {
		dummyOdds.next = evens
	} else {
		dummyOdds.next.next = evens
	}

	head = odds

	return head
}

export { SinglyLinkedList }
